pkg/kstatus/README.md

```diff
@@ -58,6 +58,25 @@ solution to this issue is to adopt the pattern used by several of the built-in t
 If the `generation` and the `observedGeneration` of a resource does not match, it means there are changes
 that the controller has not yet seen, and therefore not acted upon.
 
+#### The `Ready` Condition
+For resources that have not adopted the recommended conditions described above
+and we don't have a type-specific rule for the resource, the library
+will look for the `Ready` condition. If there is a `Ready` condition and it
+is `True`, the library will consider the resource to be fully reconciled. If
+there is a `Ready` condition and it is `False`, the library will consider the
+resource to be in the process of reconciling.
+
+There are a few corner cases here:
+ * If a resource doesn't set the Ready condition until it is True,
+the library have no way of telling whether the resource is using the
```

Comment:
I like the suggestion here, although I don't know if we can do this in practice. Using the `Ready` condition is widely used, while I don't think `observedGeneration` is used all that much. 
For the `Ready` condition to work reliably, we need the CRD and controller to both use the `observedGeneration` and set the condition to some value the first time the controller observe the resource. Both of these are things that should be pretty easy to add to existing CRDs without breaking any kind of backwards compatibility.

Reply:
Very true. I wish the guidance were stronger on these best practices. Right now even if the controller author digs up Api Conventions, all it says is:

> Some resources report the observedGeneration, which is the generation most recently observed by the component responsible for acting upon changes to the desired state of the resource. This can be used, for instance, to ensure that the reported status reflects the most recent desired status.

Doesn't exactly drive home how important this can be for the reliability of downstream tooling, and I don't see anything about initial condition values. Maybe the section on CRs under "Concepts" above could mention that these two things are also required for best results?

One thing we could do to mitigate the situation where a CR controller uses but does not initially set the Ready condition would be to check for it in the validation schema on the CRD. That would also be a way to check for other standard-ish conditions. That said, the complexity that would add probably isn't worth it, given that (1) such schema are optional, (2) just because the condition is valid doesn't mean the controller actually does set it and (3) once we need this, we already know we're in an edge case where the controller doesn't follow best practices. 🤔

Reply:
Yeah, checking the schema could be interesting. That would allow us to check whether a resource type use the `observedGeneration` pattern. It looks like the `observedGeneration` field on the built-in types are optional and `omitempty`, so just checking for the presence of the field on an unstructured resource will not work.

The actual list of condition types is not part of the schema as far as I can tell, so we won't be able to check if a CRD controller use a particular condition type by looking at the schema.